README.md
# TODOlistNodeJs
A simple TODO list APP using NodeJS, MongoDB and Bulma for style 


bulma-0.8.0/CHANGELOG.md
# Bulma Changelog

## 0.8.0

### Big update

#### Larger form controls

Controls and buttons are now `2.5em` high. You can revert this resizing by setting these previous values:

```sass
$control-height: 2.25em
$control-padding-vertical: calc(0.375em - #{$control-border-width})
$control-padding-horizontal: calc(0.625em - #{$control-border-width})
$button-padding-vertical: calc(0.375em - #{$button-border-width})
$button-padding-horizontal: 0.75em 
```

#### Light and dark colors

Each main color (`"primary"`, `"info"`, `"success"`, `"warning"`, `"danger"`) now has a `*-light` and `*-dark` version. They are calculated using 2 new color functions:

* `findLightColor()` which finds the light version of a color
* `findDarkolor()` which finds the dark version of a color

The light colors are used by the `button` element, while the light and dark colors are used by the `message` component.

#### Panel colors

The `panel` component is now available in all the different colors.

#### 4-value color map

The `$colors` Sass map now accepts, for each of its values, a map of up to **4** values. For example: the key `"info"` now has the `($info, $info-invert, $info-light, $info-dark)` map.

If you provide a `$custom-colors` map, you can decide to provide a map of 1, 2, 3 or 4 values for each value. If fewer than 4 are provided, Bulma will calculate the remaining ones:

```scss
$custom-colors: (
  "lime": (lime),
  "tomato": (tomato, white),
  "orange": ($orange, $orange-invert, $orange-light),
  "lavender": ($lavender, $lavender-invert, $lavender-light, $lavender-dark)
);
```

This is processed by the updated `mergeColorMaps()` Sass function.

#### Scheme variables

There are 6 new `$scheme` derived variables: `$scheme-main` `$scheme-main-bis` `$scheme-main-ter` `$scheme-invert` `$scheme-invert-bis` `$scheme-invert-ter`
They replace the `$white` and `$black` occurences in the codebase. This makes it easy to create a "Dark mode" simply by swapping the values:

```sass
$scheme-main: $black
$scheme-invert: $white
// etc.
```

That is also why most of the codebase now references **derived** variables (`$text`, `$background`, `$border` etc.) instead of **initial** ones (`$grey`, `$grey-lighter`, `$grey-darker` etc.): updating the derived variables will affect all elements and components directly.

#### Initial variables

* `$green: hsl(141, 53%, 53%)`
* `$cyan: hsl(204, 71%, 53%)`
* `$red: hsl(348, 86%, 61%)`

#### Derived variables

* `$primary-invert: findColorInvert($primary)`
* `$primary-light: findLightColor($primary)`
* `$primary-dark: findDarkColor($primary)`
* `$info-invert: findColorInvert($info)`
* `$info-light: findLightColor($info)`
* `$info-dark: findDarkColor($info)`
* `$success-invert: findColorInvert($success)`
* `$success-light: findLightColor($success)`
* `$success-dark: findDarkColor($success)`
* `$warning-invert: findColorInvert($warning)`
* `$warning-light: findLightColor($warning)`
* `$warning-dark: findDarkColor($warning)`
* `$danger-invert: findColorInvert($danger)`
* `$danger-light: findLightColor($danger)`
* `$danger-dark: findDarkColor($danger)`
* `$light-invert: findColorInvert($light)`
* `$dark-invert: findColorInvert($dark)`

* `$scheme-main: $white`
* `$scheme-main-bis: $white-bis`
* `$scheme-main-ter: $white-ter`
* `$scheme-invert: $black`
* `$scheme-invert-bis: $black-bis`
* `$scheme-invert-ter: $black-ter`

### Other variables

* `$control-height: 2.5em`
* `$control-padding-vertical: calc(0.5em - #{$control-border-width})`
* `$control-padding-horizontal: calc(0.75em - #{$control-border-width})`
* `$media-border-color: rgba($border, 0.5)`
* `$notification-code-background-color: $scheme-main`
* `$panel-radius: $radius-large`
* `$panel-shadow: 0 0.5em 1em -0.125em rgba($scheme-invert, 0.1), 0 0px 0 1px rgba($scheme-invert, 0.02)`
* `$textarea-padding: $control-padding-horizontal`
* `$textarea-max-height: 40em`
* `$textarea-min-height: 8em`

### Bug fixes

* Fix #2647 -> Missing meta tags in snippet
* Fix #2031, Fix #2483 -> Invalid output when declaring a custom shade map
* Fix #2060 -> `height: auto` on HTML `audio` element breaks height of element
* Fix #706 -> Derive `-invert` variables using `findColorInvert()`
* #1608 Fix #1552 -> `.container.is-fluid` margins

### New features

* #2563 `.image` has a new `.is-fullwidth` modifier

## 0.7.5

### Deprecation warning

The `form.sass` file is **deprecated**. It has moved into its own `/form` folder. If you were importing `form.sass`, please import `sass/form/_all.sass` now.
If you were simply importing the whole of Bulma with `@import "~/bulma/bulma.sass"` or similar, you won't have to change anything, and everything will work as bbefore.

### New features

#### Support for overriding the `font-family`

You can now specify a different `font-family` for the `.title`, `.subtitle` and `.button` by using the variables `$title-family`, `$subtitle-family` and `$button-family` respectively.

Simply set a value when importing Bulma:

```scss
$title-family: "Georgia", serif;
```

* #2375 Add `.is-relative` helper
* #2321 Make `.navbar` focus behave like hover for the navigation
* #2290 Fix #1186 -> Reset the offset on columns
* #2231 Add `.has-text-weight-medium` helper
* #2224 Add customizable border radius to progress bar
* #2480 Add `$footer-color` variable

### Improvements

* #2396 Update docs with webpack 4 example
* #2381 Make centered buttons have equal margin
* Fix #2297 -> Remove `.container` fixed width values, use `flex-grow`
* #2478 Move form.sass into its own folder

### Bug fixes

* #2420 Fix #2414 -> Fix `align` attribute in `td/th` being ignored
* #2463 Remove duplicate `.has-addons` in `tag.sass`
* #2253 Fix `$gap` variable default value
* #2273 Fix #2258 -> Fix Indeterminate Progress Bar animation in Firefox

app.js
const express = require('express');
const path = require('path')

const checkListRouter = require ('./src/routes/checklist');
const taskRouter = require ('./src/routes/task');
const rootRouter = require ('./src/routes/index');

const methodOverride = require('method-override')

require('./config/database');

const app = express();
app.use(express.json());
app.use(express.urlencoded({ extended: true }));
app.use(methodOverride('_method', {methods: ['POST', 'GET']}));

app.use(express.static(path.join(__dirname, 'public')));

app.set('views', path.join(__dirname, 'src/views'));
app.set('view engine', 'ejs');

app.use('/', rootRouter);
app.use('/checklists', checkListRouter);
app.use('/checklists', taskRouter.checklistDependent);
app.use('/tasks', taskRouter.simple);

app.listen(3000, () => {
  console.log('Servidor iniciado');
})

src/routes/task.js
const express = require("express");
const checklistDependentRoute = express.Router();
const simpleRouter = express.Router();

const Checklist = require("../models/checklist");
const Task = require("../models/task");

checklistDependentRoute.get('/:id/tasks/new', async (req, res) =>{
    try {
        const task = Task();
        res.status(200).render('tasks/new', { checklistId: req.params.id, task: task})
    } catch (error) {
      res.status(422).render('pages.error', { error: 'Erro ao carregar o formulário.'})  
    }
})

simpleRouter.delete('/:id', async (req, res) =>{
    try {
        const task = await Task.findByIdAndDelete(req.params.id);
        const checklist = await Checklist.findById(task.checklist);
        const taskToRemove = checklist.tasks.indexOf(task._id);
        checklist.tasks.splice(taskToRemove, 1);
        checklist.save()
        res.redirect(`/checklists/${checklist._id}`);
    } catch (error) {
        res.status(422).render('pages.error', { error: 'Erro ao remover uma tarefa.'})
    }
})

checklistDependentRoute.post('/:id/tasks', async (req, res) =>{
    const { name } = req.body.task;
    const task = new Task({ name, checklist: req.params.id })
    try {
        await task.save();
        const checklist = await Checklist.findById(req.params.id);
        checklist.tasks.push(task);
        await checklist.save();
        res.redirect(`/checklists/${req.params.id}`);
    } catch (error) {
        const errors = error.errors;
      res.status(422).render('tasks/new', { task: {...task, errors}, checklistId: req.params.id})  
    }
})

simpleRouter.put('/:id', async (req, res) =>{
    const task = await Task.findById(req.params.id);
    try {
        task.set(req.body.task);
        await task.save();
        res.status(200).json({ task });
    } catch (error) {
        const errors = error.errors;
        res.status(422).json({ task: {...errors }});
    }
})

module.exports = { checklistDependent: checklistDependentRoute, simple: simpleRouter }

src/routes/checklist.js
const express = require("express");
const router = express.Router();

const Checklist = require("../models/checklist");

router.get("/", async (req, res) => {
  try {
    const checklist = await Checklist.find({});
    res.status(200).render('checklists/index', { checklist : checklist})
  } catch (error) {
    res.status(500).render('pages/error', { error: 'Erro ao exibir as Listas'});
  }
});

router.get('/new', async (req, res) =>{
try {
  const checklist = new Checklist();
  res.status(200).render('checklists/new', { checklist: checklist})
} catch (error) {
  res.status(500).render('pages/error', {error: 'Erro ao carregar o formulário.'})
}
})

router.get('/:id/edit', async (req, res) =>{
  try {
    const checklist = await Checklist.findById(req.params.id);
    res.status(200).render('checklists/edit', { checklist: checklist});
  } catch (error) {
    res.status(500).render('pages/error', {error: 'Erro ao exibir a edição de lista de tarefas.'})
  }
})

router.post("/", async (req, res) => {
  const { name } = req.body.checklist;
  const checklist = new Checklist({ name })
  try {
    await checklist.save();
    res.redirect('/checklists')
  } catch (error) {
    res.status(422).render('checklists/new', { checklist: {...checklist, error}});
  }
});

router.get("/:id", async (req, res) => {
  try {
    const checklist = await Checklist.findById(req.params.id).populate('tasks');
    res.status(200).render('checklists/show', { checklist: checklist})
  } catch (error) {
    res.status(500).render('pages/error', { error: 'Erro ao exibir as listas de tarefas'});
  }
});

router.put("/:id", async (req, res) => {
  const { name } = req.body.checklist;
  const checklist = await Checklist.findById(req.params.id);

  try {
    await checklist.updateOne({ name });
    res.redirect('/checklists');
  } catch (error) {
    const errors = error.erros;
    res.status(422).render('checklists/edit', { checklist: {...checklist, errors}});
  }
});

router.delete("/:id", async (req, res) => {
  try {
    await Checklist.findByIdAndRemove(req.params.id);
    res.redirect('/checklists')
  } catch (error) {
    res.status(500).render('pages/error', { error: 'Erro ao deletar a lista de tarefas'});
  }
});

module.exports = router;